lightning/src/ln/functional_test_utils.rs

```diff
@@ -2774,7 +2789,11 @@ pub fn do_pass_along_path<'a, 'b, 'c>(args: PassAlongPathArgs) -> Option<Event>
 				}
 				event = Some(events_2[0].clone());
 			} else if let Some(ref failure) = expected_failure {
-				assert_eq!(events_2.len(), 2);
+				// If we successfully decode the HTLC onion but then fail later in
+				// process_pending_htlc_forwards, then we'll queue the failure and generate a new
+				// `ProcessPendingHTLCForwards` event. If we fail during the process of decoding the HTLC,
+				// we'll fail it immediately with no intermediate forwarding event.
+				assert!(events_2.len() == 1 || events_2.len() == 2);
```

Comment:
Wouldn't this pass when two events is expected but a regression results in only one? Is there another way to assert that would avoid this? Maybe through what's in `expected_failure` or in how it set?

Reply:
IIRC I couldn't find a way to tell with the current code. This would be fixed if https://github.com/lightningdevkit/rust-lightning/issues/1302#issuecomment-2593864258 (or possibly https://github.com/lightningdevkit/rust-lightning/issues/3561) were addressed, though. There really shouldn't be 1 event ever except for the bug documented in the former link.